File: lib/features/ccmoon/auth_token.dart
import 'package:dio/dio.dart';

const _authUrl = 'https://slbsso.meijo-u.ac.jp/opensso/json/authenticate';

Future<String> getToken(
  String userid,
  String password, {
  Dio? dio,
  int maxAttempts = 20,
  Duration retryInterval = const Duration(milliseconds: 500),
}) async {
  final client = dio ?? Dio();
  final headers = {'Content-Type': 'application/json'};

  try {
    final initial = await client.post(
      _authUrl,
      options: Options(headers: headers),
    );
    final body = Map<String, dynamic>.from(initial.data as Map);
    body['callbacks'][0]['input'][0]['value'] = userid;
    body['callbacks'][1]['input'][0]['value'] = password;

    Response? token;
    for (var i = 0; i < maxAttempts; i++) {
      token = await client.post(
        _authUrl,
        data: body,
        options: Options(
          headers: headers,
          validateStatus: (_) => true,
        ),
      );
      if (token.statusCode == 200) break;
      await Future.delayed(retryInterval);
    }
    if (token == null || token.statusCode != 200) {
      throw StateError('認証エンドポイント応答異常 (last=${token?.statusCode})');
    }
    return (token.data as Map)['tokenId'] as String;
  } catch (e, st) {
    Error.throwWithStackTrace(
      Exception('tokenを取得することができませんでした.時間をおいて再度試してください. ($e)'),
      st,
    );
  }
}

File: features/ccmoon/auth_token.py
import requests
import time
import json

def getToken(userid,password):
    try:
        url = 'https://slbsso.meijo-u.ac.jp/opensso/json/authenticate'
        
        headers = {
            'Content-Type' : 'application/json'
        }
        source = requests.post(url,headers=headers)
        jsn = json.loads(source.text)
        
        jsn["callbacks"][0]["input"][0]["value"] = userid
        jsn["callbacks"][1]["input"][0]["value"] = password
        statuscode = 0
        for i in range(20):
            token = requests.post(url,headers=headers,json=jsn)
            statuscode = token.status_code
            if statuscode == 200:
                break
            time.sleep(0.5)
            
        succesURL = json.loads(token.text)
        tokenId = succesURL["tokenId"]
        #kugiri()
        return tokenId
    except:
        raise Exception("tokenを取得することができませんでした.時間をおいて再度試してください.")